Update ORDER BY section

sazar1 opened the pull request:
Following the discussion here: 
https://firebolt-analytics.slack.com/archives/C02SMTT26VA/p1709555190448469

sazar1 committed:
Update select.md

Following the discussion here: 
https://firebolt-analytics.slack.com/archives/C02SMTT26VA/p1709555190448469

ahopp commented on docs/sql-reference/commands/select.md:
```diff
@@ -661,7 +661,7 @@ When including multiple clauses, the same number of columns must be selected by
 The `ORDER BY` clause sorts a result set by one or more output expressions. `ORDER BY` is evaluated as the last step after any `GROUP BY` or `HAVING` clause. `ASC` and `DESC` determine whether results are sorted in ascending or descending order. When the clause contains multiple expressions, the result set is sorted according to the first expression. Then the second expression is applied to rows  ...
 
-The default null ordering is `NULLS LAST`, regardless of ascending or descending sort order.
+By default, NULL values sort as if larger than any non-null value; that is, NULLS FIRST is the default for DESC order, and NULLS LAST otherwise.
```
I know this is pulled from the postgres [documentation](https://www.postgresql.org/docs/current/queries-order.html#:~:text=By%20default%2C%20null%20values%20sort,independently%20for%20each%20sort%20column.) but I think we can improve the legibility a bit by speaking or order (e.g., first, last) versus mechanism (e.g., larger, smaller). We might also want to add the ASC behavior explicitly like the postgres docs?

Perhaps something like this;

> By default, NULL values are sorted ahead of non-NULL values ('NULL FIRST') unless otherwise specified. With 'NULL LAST', NULL values are sorted after non-NULL values. 

NOTE: I'm a bit confused on the default order, so we should double check the functionality follows our documented behavior either way. Specifically, from the postgres docs, if a user specifies `ORDER BY`, `ASC` order is the default thus NULL values will be sorted `ASC` - and since `NULLS LAST` is the non-`DESC` default I'd assume NULLS would be sorted last by default BUT we also have "By default, null values sort as if larger than any non-null value" which seems to imply the opposite.

sazar1 replied:
Thanks for taking a look on this one Andrew. 
I agree with the suggested draft. 

As mentioned [here](https://firebolt-analytics.slack.com/archives/C02SMTT26VA/p1709561500187969?thread_ts=1709555190.448469&cid=C02SMTT26VA) about this behavior:
> It's the default in LQP2, no idea about LQP1. But it will definitely be the default in FB 2.0 

Adding @leonard-firebolt, @lorenzhs if there are additional insights.

leonard-firebolt replied:
Our behavior matches PostgreSQL's behavior. Here is an example, which uses the format of our SQL test framework:
```
SQL command
--
expected result
==
```
Example:
```
CREATE TABLE test (a INT NULL);
--
OK
==
INSERT INTO test VALUES (1), (2), (NULL);
--
OK
==
SELECT a FROM test ORDER BY a;
--
a INTEGER
1
2
NULL
==
SELECT a FROM test ORDER BY a NULLS FIRST;
--
a INTEGER
NULL
1
2
==
SELECT a FROM test ORDER BY a NULLS LAST;
--
a INTEGER
1
2
NULL
==
SELECT a FROM test ORDER BY a ASC;
--
a INTEGER
1
2
NULL
==
SELECT a FROM test ORDER BY a ASC NULLS FIRST;
--
a INTEGER
NULL
1
2
==
SELECT a FROM test ORDER BY a ASC NULLS LAST;
--
a INTEGER
1
2
NULL
==
SELECT a FROM test ORDER BY a DESC;
--
a INTEGER
NULL
2
1
==
SELECT a FROM test ORDER BY a DESC NULLS FIRST;
--
a INTEGER
NULL
2
1
==
SELECT a FROM test ORDER BY a DESC NULLS LAST;
--
a INTEGER
2
1
NULL

```
Here is proof that that's the PostgreSQL result (run against PostgreSQL using our `PostgresExecutor`):
```
./tests/sql/sql_test --gtest_filter='*leonard*' --postgres
Note: Google Test filter = *leonard*
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from LocalExecutorQueries/SqlTest
[ RUN      ] LocalExecutorQueries/SqlTest.RunTestFile/leonard
[       OK ] LocalExecutorQueries/SqlTest.RunTestFile/leonard (42 ms)
[----------] 1 test from LocalExecutorQueries/SqlTest (42 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (1094 ms total)
[  PASSED  ] 1 test.
```

leonard-firebolt replied:
The behavior documented in PG matches our behavior and is IMO descriptive and precise enough. I would recommend to copy it and slightly change it to:

> The `NULLS FIRST` and `NULLS LAST` options can be used to determine whether nulls appear before or after non-null values in the sort ordering. By default, null values sort as if larger than any non-null value; that is, `NULLS FIRST` is the default for `DESC` order, and `NULLS LAST` is the default for `ASC` order.

Note that we will build a new, faster sort operator starting next month and we will revisit the documentation then again.

leonard-firebolt replied:
Speaking of "larger" (defined in terms of the > operator) and "smaller" (defined in terms of the < operator) is IMO a very unambiguous way to describe sort order.

ahopp replied:
> a very unambiguous way to describe sort order

Fair, @leonard-firebolt, but I'm not sure I agree its the _most_ clear (primarily since "order", and not relative size, seems to be the intuitive difference in operators). BUT I'll trust your judgement here - you and what @sazar1 decides.

ahopp replied:
> Note that we will build a new, faster sort operator starting next month and we will revisit the documentation then again.

Ack!

ahopp reviewed (commented):
Just a quick comment.

ahopp reviewed (commented):
@sazar1 I'll trust you and @leonard-firebolt to make the call - either way framing works for me.

sazar1 committed:
Update select.md

Edited the suggested phrasing as discussed here: 
https://firebolt-analytics.slack.com/archives/C016HSVDP9U/p1709660686592699?thread_ts=1709567260.704049&cid=C016HSVDP9U

ahopp commented:
@sazar1 I don't have approving privileges to this repo but I say ship it!  

Thanks @sazar1 and @leonard-firebolt!
